runtime/lua/vim/diagnostic.lua

```diff
@@ -377,6 +377,37 @@ local function clear_scheduled_display(namespace, bufnr)
   end
 end
 
+---@private
+---@param severity lsp.DiagnosticSeverity
+local function severity_vim_to_lsp(severity)
+  if type(severity) == 'string' then
+    severity = vim.diagnostic.severity[severity]
+  end
+  return severity
+end
+
+--- @private
+--- @param diagnostic Diagnostic
+--- @return lsp.Diagnostic[]
+local function diagnostic_vim_to_lsp(diagnostic)
+  return vim.tbl_extend('keep', {
+    range = {
+      start = {
+        line = diagnostic.lnum,
+        character = diagnostic.col,
+      },
+      ['end'] = {
+        line = diagnostic.end_lnum,
+        character = diagnostic.end_col,
+      },
+    },
+    severity = severity_vim_to_lsp(diagnostic.severity),
+    message = diagnostic.message,
+    source = diagnostic.source,
+    code = diagnostic.code,
+  }, diagnostic.user_data and (diagnostic.user_data.lsp or {}) or {})
+end
+
```

Comment:
`vim.diagnostic` should not have any LSP-isms in it. The `vim.lsp.diagnostic` module serves as the "bridge" between "LSP diagnostics" and the generic `vim.diagnostic` interface. We should not add these here.

Reply:
maybe can move to util?

Reply:
moved it to util.lua and this function could still use some improvements. Implement a two-way conversion from diagnostics to vim and from vim to lsp. This can reduce some private functions in `vim.lsp.diagnsotic`. what do you think ?